apps/prairielearn/src/lib/course.ts

```diff
@@ -0,0 +1,197 @@
+import { z } from 'zod';
+import { callbackify } from 'util';
+import * as fs from 'fs-extra';
+import * as namedLocks from '@prairielearn/named-locks';
+import * as sqldb from '@prairielearn/postgres';
+
+import { createServerJob } from './server-jobs';
+import { config } from './config';
+import * as chunks from './chunks';
+import { syncDiskToSqlWithLock } from '../sync/syncFromDisk';
+import { IdSchema } from './db-types';
+import {
+  getCommitHashAsync,
+  getOrUpdateCourseCommitHashAsync,
+  updateCourseCommitHashAsync,
+} from './courseUtil';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+/**
+ * Check that an assessment_instance_id really belongs to the given course_instance_id
+ *
+ * @param assessment_instance_id - The assessment instance to check.
+ * @param course_instance_id - The course instance it should belong to.
+ */
+export async function checkBelongsAsync(
+  assessment_instance_id: string,
+  course_instance_id: string,
+): Promise<void> {
+  if (
+    sqldb.queryOptionalRow(
+      sql.check_belongs,
+      {
+        assessment_instance_id,
+        course_instance_id,
+      },
+      IdSchema,
+    ) == null
+  ) {
+    throw new Error('access denied');
+  }
+}
+export const checkBelongs = callbackify(checkBelongsAsync);
+
+/**
+ * Return the name and UID (email) for every owner of the specified course.
+ *
+ * @param {string} course_id The ID of the course.
+ * @returns {Promise<{ uid: string, name?: string }[]>}
+ */
+export async function getCourseOwners(
+  course_id: string,
+): Promise<{ uid: string; name?: string }[]> {
+  const { rows } = await sqldb.queryAsync(sql.select_owners, { course_id });
+  return rows.map((row) => ({
+    uid: row.uid,
+    name: row.name,
+  }));
+}
+
+export function getLockNameForCoursePath(coursePath) {
+  return `coursedir:${coursePath}`;
+}
+
+const CourseDataSchema = z.object({
+  path: z.string().nullable(),
+  branch: z.string().nullable(),
+  repository: z.string().nullable(),
+  commit_hash: z.string().nullable(),
+});
+
+export async function pullAndUpdate({
```

Comment:
Initial plan was to have this in `courseUtil.ts`, but that generated a circular dependency (courseUtil -> syncFromDisk -> freeform server -> courseUtil) which was causing problems when running tests.